libraries/rush-sdk/src/proxy.ts

```diff
@@ -0,0 +1,288 @@
+    ___rush___rushLibModule?: RushLibModuleType;
+    ___rush___rushLibModuleFromEnvironment?: RushLibModuleType;
+    ___rush___rushLibModuleFromInstallAndRunRush?: RushLibModuleType;
+  };
+
+/**
+ * Type of {@link ISdkCallbackEvent.logMessage}
+ * @public
+ */
+export interface IProgressBarCallbackLogMessage {
+  /**
+   * A status message to print in the log window, or `undefined` if there are
+   * no further messages.  This string may contain newlines.
+   */
+  message: string;
+
+  /**
+   * The type of message.  More message types may be added in the future.
+   */
+  messageType: 'info' | 'debug';
+}
+
+/**
+ * Type of {@link ISdkCallbackEvent.state}
+ * @public
+ */
+export type SdkCallbackState =
+  /**
+   * The task has started and is still running.
+   */
+  | 'running'
+  /**
+   * The task has completed successfully. No further events will be fired.
+   */
+  | 'succeeded'
+  /**
+   * The task was aborted by the caller. No further events will be fired.
+   */
+  | 'aborted'
+  /**
+   * The task has failed due to an error. No further events will be fired.
+   */
+  | 'failed';
+
+/**
+ * Event options for {@link ILoadSdkAsyncOptions.onNotifyEvent}
+ * @public
+ */
+export interface ISdkCallbackEvent {
+  /**
+   * Indicates the current state of the operation.
+   */
+  state: SdkCallbackState;
+
+  /**
+   * Allows the caller to display log information about the operation.
+   */
+  logMessage: IProgressBarCallbackLogMessage | undefined;
+
+  /**
+   * Allows the caller to display a progress bar for long-running operations.
+   *
+   * @remarks
+   * If a long-running operation is required, then `progressBarPercent` will
+   * start at 0.0 and count upwards and finish at 100.0 if the operation completes
+   * successfully.  If the long-running operation has not yet started, or
+   * is not required, then the value will be `undefined`.
+   */
+  progressBarPercent: number | undefined;
+}
+
+/**
+ * Type of {@link ILoadSdkAsyncOptions.onNotifyEvent}
+ * @public
+ */
+export type SdkNotifyEventCallback = (sdkEvent: ISdkCallbackEvent) => void;
+
+/**
+ * Options for {@link rushSdkProxy.loadAsync}
+ * @public
+ */
+export interface ILoadSdkAsyncOptions {
+  /**
+   * The folder to start from when searching for the Rush workspace configuration.
+   * If this folder does not contain a `rush.json` file, then each parent folder
+   * will be searched.  If `rush.json` is not found, then the SDK fails to load.
+   */
+  rushJsonSearchFolder?: string;
+
+  /**
+   * A cancellation token that the caller can use to prematurely abort the operation.
+   */
+  abortSignal?: AbortSignal;
+
+  /**
+   * Allows the caller to monitor the progress of the operation.
+   */
+  onNotifyEvent?: SdkNotifyEventCallback;
+}
+
+/**
+ * Exposes operations that control how the `@microsoft/rush-lib` engine is
+ * located and loaded.
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-namespace
+export namespace rushSdkProxy {
```

Comment:
Why the superfluous namespace? we can export `loadAsync` directly from the module.

Reply:
We've chatted about this before. I think it's not a good design for public APIs to have loose functions.  The name of this function is `rushSdkProxy.loadAsync`, and our ESLint rules should enforce that it is always written out like that. It should not be written as `loadAsync` because that is a highly ambiguous name.

Reply:
Strictly speaking, the name of the function would be `{ loadAsync } from '@rushstack/rush-sdk/proxy'`, which is already quite specific. Alternatively, it could be `{ loadRushLibAsync } from '@rushstack/rush-sdk/proxy'`, which I would find preferable to destroying the compressiblity of bundling `rush-sdk` by wrapping it in a namespace object.